kernel/src/snake.rs

```diff
@@ -0,0 +1,366 @@
+use core::sync::atomic::{Ordering, AtomicU64};
+use alloc::vec::Vec;
+use crate::terminal::{TerminalOutput, TerminalCharacter, Point, STDOUT};
+use crate::drivers::{pit, ps2};
+use crate::drivers::keyboard::{Ps2Keyboard, Keyboard, KeyEventType, keymap::codes::*};
+use crate::halt;
+
+const HEAD_CHAR: char = 2 as char;
+lazy_static! {
+    static ref RNG: Random = Random::new();
+}
+
+struct Game<'a> {
+    grid: Grid,
+    snake: Snake,
+    ups: usize,
+    keyboard: &'a mut Ps2Keyboard<'a>,
+    highscore: u16,
+}
+
+impl<'a> Game<'a> {
+    fn new(keyboard: &'a mut Ps2Keyboard<'a>) -> Game<'a> {
+        let res = STDOUT.read().resolution();
+
+        Game {
+            grid: Grid::empty(res.x as usize, res.y as usize),
+            snake: Snake::new(),
+            ups: 20,
+            keyboard,
+            highscore: 0,
+        }
+    }
+
+    fn run(&mut self) {
+        STDOUT.write().clear().expect("Error clearing screen");
+        self.notification("Welcome to snake! Press any key to continue...");
+
+        STDOUT.write().clear().expect("Error clearing screen");
+        self.grid.set(generate_apple_pos(&self.grid), Cell::Apple);
```

Comment:
Maybe this should call `restart` such that if we change any initialisation behaviour it is consistent?

Reply:
Restart will print a win/loss message though.